pkg/operator/controller/ingress/load_balancer_service.go

```diff
@@ -159,3 +163,21 @@ func (r *reconciler) finalizeLoadBalancerService(ci *operatorv1.IngressControlle
 	}
 	return nil
 }
+
+// ensureLoadBalancerCleanupFinalizer ensures the "service.kubernetes.io/load-balancer-cleanup"
+// finalizer does not exist for the load balancer service of ic. For additional background on
+// this finalizer, see:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190423-service-lb-finalizer.md
+func (r *reconciler) ensureLoadBalancerCleanupFinalizer(ic *operatorv1.IngressController) error {
+	service, err := r.currentLoadBalancerService(ic)
+	if err != nil {
+		return err
+	}
+	if service == nil {
+		return nil
+	}
+	if slice.ContainsString(service.Finalizers, lbServiceCleanUpFinalizer) {
```

Comment:
Why return nil if the service exists but does not have the finalizer?

Reply:
@Miciah I pushed a new commit to address your feedback.

Reply:
I mean, why shouldn't `ensureLoadBalancerCleanupFinalizer` always return an error if the service exists?

Reply:
I was thinking the func can be used for other purposes in the future where we expect the service to exist but not the finalizer, i.e. bare metal providers.

Reply:
Just for clarity, using LB Service publishing is the only time we create a service at all today, and the only new one we've talked about is NodePort.

Reply:
If the service exists without a finalizer, why is it better to delete the ingresscontroller immediately rather than waiting till the service has been deleted?

Reply:
@Miciah I was not necessarily thinking the func had to be used for a delete operation. I'll update the func based on your initial feedback.

Reply:
I'm still not getting it.  Maybe it would be useful to take another resource as an example: In the case of the DNSRecord resource, we return before removing finalizer from the IngressController resource if the DNSRecord is still present:  https://github.com/openshift/cluster-ingress-operator/blob/a7f75ef84133e0ae1f8cd71afec2517becfa5099/pkg/operator/controller/ingress/controller.go#L474-L479

This way, we delay the actual deletion of an ingresscontroller as long as a corresponding DNSRecord resource still exists.  This way, the ingresscontroller cannot be deleted (and subsequently recreated) while that old DNSRecord still exists.

It seems to me like we should do the same thing with the Service resource:  Return before removing the finalizer from the IngressController resource until the corresponding service is gone, because we do not want to risk that a new ingresscontroller gets created while the old one's Service resource still exists.  I don't see that it matters what finalizers are on a Service resource—if a service exists, we should keep the finalizer on the IngressController resource so that it cannot be deleted (and subsequently recreated) while the old service still exists.

Is there a situation where we should allow deletion of the ingresscontroller while the service still exists?

Reply:
Seems like blockOwnerDeletion=true neatly solves this, if only we had the right "anchor resource" in the openshift-ingress namespace. Then we could constrain all the "waiting" logic to just the anchor object rather than code to wait for every dependent.

Reply:
Our anchor in the `openshift-ingress` namespace is the Deployment resource.

Reply:
https://github.com/openshift/cluster-ingress-operator/pull/320/files#r341771387 makes sense to me. Then we don't need to know anything about the upstream finalizer at all.

@danehans I think this would be even simpler...